### DISCUSSION TITLE
How to compare MiniDexed properly to a DX1 or DX5; Performances

### DISCUSSION THREAD

**Discussion by probonopd** [probonopd/MiniDexed#554] [General]
Rob "FM" Puricelli ([characterized](https://www.soundonsound.com/author/rob-puricelli) by SOS as "a Music Technologist and Instructional Designer who has a healthy obsession with classic synthesizers and their history") is [considering](https://www.youtube.com/live/TDSy5nnm0jA?feature=shared&t=10323) to do a comparison between MiniDexed and... the real thing:

> "I'm actually tempted to do a side-by-side."

Wait... did he actually mean... these guys from his [website](https://www.failedmuso.com/)? :eyes:

[![image](https://github.com/probonopd/MiniDexed/assets/2480569/a64f3610-2bea-4ada-8380-603217e63be8)](https://www.failedmuso.com/wp-content/uploads/2023/01/324234574_693242178868728_2456946444318374534_n.jpg)

Only 140 DX1 were ever made, and even though it is approaching its 40th anniversary it is arguably still the finest FM instrument to this date, with 2 tone generators, 73-key weighted wooden keyboard with polyphonic aftertouch, handmade Brazilian rosewood case, and... polyphonic aftertouch.

How would one even go about comparing MiniDexed to a DX1?

* A properly jumpered [DAC](https://github.com/probonopd/MiniDexed/wiki/Hardware#i2c-dac) should be used
* Possibly `EngineType=2` should be used instead of the default ([details](https://github.com/probonopd/MiniDexed/wiki/Files#sound-device))
* The DX1/DX5 voices are listed on page 8 of the [DX5 manual](https://www.synthxl.com/wp-content/uploads/2020/06/Yamaha-DX-5-Owner-Manual.pdf). http://hsjp.eu/downloads/Dexed/Dexed_cart_1.0.zip contains `DX1A1.SYX`, `DX1A2.SYX`, `DX1B1.SYX`, `DX1B2.SYX` which should run fine on MiniDexed
* The DX1 and DX5 have factory 64 preprogrammed performance combinations, which was also delivered on a performance ROM cartridge. So far I didn't think we had access to the data from the performance cartridge and could only guess how the performances are configured based on the rather vague descriptions on page 8 of the [DX5 manual](https://www.synthxl.com/wp-content/uploads/2020/06/Yamaha-DX-5-Owner-Manual.pdf). But look who had [successfully](https://yamahamusicians.com/forum/viewtopic.php?t=14512) looked for those before? Right... the very Rob "FM" Puricelli. Which brings us to the question... how to interpret the DX1/DX5 performance sysex, so that we could configure MiniDexed in a similar way...
* For the comparison, we need to think about the maximum MIDI value 100 vs 127, depending on which master keyboard is used for the comparison the numbers coming from the keyboard must be interpreted differently (in order not to fall into [this](https://www.youtube.com/watch?v=iHycGwuWcn0) trap)

**Comment by probonopd** [comment]
> Which brings us to the question... how to interpret the DX1/DX5 performance sysex

Page 47 of the [DX5 manual](https://www.synthxl.com/wp-content/uploads/2020/06/Yamaha-DX-5-Owner-Manual.pdf) has this to say:

![image](https://github.com/probonopd/MiniDexed/assets/2480569/be0bdb27-d2cc-494b-b4c1-116804cca293)

Maybe this can guide us in interpreting the `DX5 Performance.syx` file?

[DX5 Carts.zip](https://github.com/probonopd/MiniDexed/files/13230903/DX5.Carts.zip)

[`dxconvert.py`](https://github.com/rarepixel/dxconvert) claims to support TX7 but not DX5 performances so far. (I couldn't get it to read the `DX5 Performance.syx` file properly.)

**Reply by probonopd** [reply]
Looks like someone (Ben Hall?) has figured out the DX1/DX5 performance data format but not documented it?
http://difficultaudio.com/sounds [contacted by mail]

**Reply by probonopd** [reply]
Is the "Function Editor" in https://squest.com/Products/MidiQuest13/Instruments/YamahaDX5/index.html dealing with DX1/DX5 performance data? Is "Function Bank and Function" = Performance?

**Reply by Banana71** [reply]
>Is the "Function Editor" in https://squest.com/Products/MidiQuest13/Instruments/YamahaDX5/index.html dealing with DX1/DX5 performance data? Is "Function Bank and Function" = Performance?

Yes, I would see it the same way. Function=Performance, Function Bank= 64 Performances

**Reply by BobanSpasic** [reply]
DX5 Performance sysex dump looks to me like TX7 dump. It has the same header signature.
TX7 Function dump has also 64 places, where TX7 uses just the first 32. It seems that the dump format is designed not just for TX7

**Comment by Banana71** [comment]
>How would one even go about comparing MiniDexed to a DX1?

First, I hope that his miniDexed has a DAC. 

>For the comparison, we need to think about the maximum MIDI value 100 vs 127, depending on which master keyboard is used for the comparison the numbers coming from the keyboard must be interpreted differently (in order not to fall into [this](https://www.youtube.com/watch?v=iHycGwuWcn0) trap)

If he uses the DX1 or DX5 as a master keyboard for the miniDexed, the sounds programmed for the Yamaha synthesizers will also sound the same.
Only the miniDexed Performances could sound a little quieter and less aggressive with a DX1 or DX5 as the master keyboard. A different master keyboard that also outputs higher velocity values would make more sense for this. But I'm sure he'll know that or figure it out for himself very quickly.
If I understood it correctly, his DX1 also makes quite a bit of noise.

>how to interpret the DX1/DX5 performance sysex, so that we could configure MiniDexed in a similar way...

@probonopd Why? Do you want to replace the Performance.ini with a performance.syx in the future?

**Reply by probonopd** [reply]
> @probonopd Why? Do you want to replace the Performance.ini with a performance.syx in the future?

You guessed my mind :+1:

**Reply by Banana71** [reply]
Where do you see the advantage?
Why replace a file that can be read and changed by the user on the PC with a scripted file that can only be changed on the PC with a miniDexed or a special editor?
I have always seen Performance.ini as an advantage.
Of course, the Performance.ini is much larger than the actual data that is needed due to the text mode, but that wouldn't make things much more complicated.
Or is it that only one Performance.syx has to be read in for a performance bank (perhaps 64 or 128 performances) and this promises you more flexibility in dealing with the performances?

**Reply by probonopd** [reply]
Well, I didn't think of "replace", rather "convert". So that one can play (something similar as) the DX1/DX5 performances on the MiniDexed by putting the same parameters as in the DX1/DX5 factory performance data into our Performance.ini. We already have access to the voice data needed, but we still need to find out the performance data.

**Reply by BobanSpasic** [reply]
I was on this way with MiniDexedCC, but I've put it on hold because nobody got interested. I already have all the data structures implemented (DX7, TX7, DX7II)

**Reply by probonopd** [reply]
Wow. So what do you think of the idea to be able to open performance.syx and save them as Performance.ini?
By the way, we should link to MiniDexedLibrarian more prominently. Will do so in the README.

**Reply by BobanSpasic** [reply]
Here are the data structures from TX7 (written in Pascal). There are a lot of "not used" bytes - it can be that these are used on DX1/DX5.

```
type
  TTX7_PMEM_Dump = array [0..63] of byte;
  TTX7_PCED_Dump = array [0..93] of byte;

type
  TTX7_PCED_Params = packed record
    case boolean of
      True: (params: TTX7_PCED_Dump);
      False: (
        A_NU0: byte;                      //       Not used
        A_NU1: byte;                      //       Not used (MIDI Channel)
        A_PolyMono: byte;                 //       0-1
        A_PitchBendRange: byte;           //       0-12
        A_PitchBendStep: byte;            //       0-12
        A_PortamentoTime: byte;           //       0-99
        A_PortaGlissando: byte;           //       0-1
        A_PortamentoMode: byte;           //       0-1
        A_NU8: byte;                      //       Not used
        A_ModWheelSens: byte;             //       0-15  translates to 0-99 for DX7
        A_ModWheelAssign: byte;           //       0-7
        A_FootCtrlSens: byte;             //       0-15  translates to 0-99 for DX7
        A_FootCtrlAssign: byte;           //       0-7
        A_AfterTouchSens: byte;           //       0-15  translates to 0-99 for DX7
        A_AfterTouchAssign: byte;         //       0-7
        A_BrthCtrlSens: byte;             //       0-15  translates to 0-99 for DX7
        A_BrthCtrlAssign: byte;           //       0-7
        A_NU17: byte;                     //       Not used
        A_NU18: byte;                     //       Not used
        A_NU19: byte;                     //       Not used
        A_NU20: byte;                     //       Not used
        A_NU21: byte;                     //       Not used
        A_NU22: byte;                     //       Not used
        A_NU23: byte;                     //       Not used
        A_NU24: byte;                     //       Not used
        A_NU25: byte;                     //       Not used
        A_VoiceAttn: byte;                //       0-7  translates to 0-99 for DX7
        A_NU27: byte;                     //       Not used
        A_NU28: byte;                     //       Not used
        A_NU29: byte;                     //       Not used

        B_NU0: byte;                      //       Not used
        B_NU1: byte;                      //       Not used
        B_PolyMono: byte;                 //       0-1
        B_PitchBendRange: byte;           //       0-12
        B_PitchBendStep: byte;            //       0-12
        B_PortamentoTime: byte;           //       0-99
        B_PortaGlissando: byte;           //       0-1
        B_PortamentoMode: byte;           //       0-1
        B_NU8: byte;                      //       Not used
        B_ModWheelSens: byte;             //       0-15  translates to 0-99 for DX7
        B_ModWheelAssign: byte;           //       0-7
        B_FootCtrlSens: byte;             //       0-15  translates to 0-99 for DX7
        B_FootCtrlAssign: byte;           //       0-7
        B_AfterTouchSens: byte;           //       0-15  translates to 0-99 for DX7
        B_AfterTouchAssign: byte;         //       0-7
        B_BrthCtrlSens: byte;             //       0-15  translates to 0-99 for DX7
        B_BrthCtrlAssign: byte;           //       0-7
        B_NU17: byte;                     //       Not used
        B_NU18: byte;                     //       Not used
        B_NU19: byte;                     //       Not used
        B_NU20: byte;                     //       Not used
        B_NU21: byte;                     //       Not used
        B_NU22: byte;                     //       Not used
        B_NU23: byte;                     //       Not used
        B_NU24: byte;                     //       Not used
        B_NU25: byte;                     //       Not used
        B_VoiceAttn: byte;                //       0-7  translates to 0-99 for DX7
        B_NU27: byte;                     //       Not used
        B_NU28: byte;                     //       Not used
        B_NU29: byte;                     //       Not used

        G_NU60: byte;                     //       Not used
        G_VoiceMemSelFlag: byte;          //       0-1
        G_NU62: byte;                     //       Not used
        G_NU63: byte;                     //       Not used
        G_PerfName01: byte;               //       ASCII
        G_PerfName02: byte;               //       ASCII
        G_PerfName03: byte;               //       ASCII
        G_PerfName04: byte;               //       ASCII
        G_PerfName05: byte;               //       ASCII
        G_PerfName06: byte;               //       ASCII
        G_PerfName07: byte;               //       ASCII
        G_PerfName08: byte;               //       ASCII
        G_PerfName09: byte;               //       ASCII
        G_PerfName10: byte;               //       ASCII
        G_PerfName11: byte;               //       ASCII
        G_PerfName12: byte;               //       ASCII
        G_PerfName13: byte;               //       ASCII
        G_PerfName14: byte;               //       ASCII
        G_PerfName15: byte;               //       ASCII
        G_PerfName16: byte;               //       ASCII
        G_PerfName17: byte;               //       ASCII
        G_PerfName18: byte;               //       ASCII
        G_PerfName19: byte;               //       ASCII
        G_PerfName20: byte;               //       ASCII
        G_PerfName21: byte;               //       ASCII
        G_PerfName22: byte;               //       ASCII
        G_PerfName23: byte;               //       ASCII
        G_PerfName24: byte;               //       ASCII
        G_PerfName25: byte;               //       ASCII
        G_PerfName26: byte;               //       ASCII
        G_PerfName27: byte;               //       ASCII
        G_PerfName28: byte;               //       ASCII
        G_PerfName29: byte;               //       ASCII
        G_PerfName30: byte;               //       ASCII
      )
  end;

  TTX7_PMEM_Params = packed record
    case boolean of
      True: (params: TTX7_PMEM_Dump);
      False: (
        A_PolyMono: byte;                 //       PM |    |    |    |    |    |    |
        A_PBSlo_PBR: byte;                //          PBS Lo    |        PBR        |
        A_PortaTime: byte;                //       0-99
        A_PM_PGL: byte;                   //       0  |                   | M  | GL |
        A_MWA_MWS: byte;                  //           MWA      |        MWS        |
        A_FCA_FCS: byte;                  //           FCA      |        FCS        |
        A_ATA_ATS: byte;                  //           ATA      |        ATS        |
        A_BCA_BCS: byte;                  //           BCA      |        BCS        |
        A_NU08: byte;
        A_NU09: byte;
        A_NU10: byte;
        A_NU11: byte;
        A_NU12: byte;
        A_NU13: byte;
        A_ATN: byte;                      //       0 |               |    ATN       |
        A_PBShi: byte;                    //       (MSB)

        B_PolyMono: byte;                 //       PM |    |    |    |    |    |    |
        B_PBSlo_PBR: byte;                //          PBS Lo    |        PBR        |
        B_PortaTime: byte;                //       0-99
        B_PM_PGL: byte;                   //       0  |                   | M  | GL |
        B_MWA_MWS: byte;                  //           MWA      |        MWS        |
        B_FCA_FCS: byte;                  //           FCA      |        FCS        |
        B_ATA_ATS: byte;                  //           ATA      |        ATS        |
        B_BCA_BCS: byte;                  //           BCA      |        BCS        |
        B_NU08: byte;
        B_NU09: byte;
        B_NU10: byte;
        B_NU11: byte;
        B_NU12: byte;
        B_NU13: byte;
        B_ATN: byte;                      //       0 |               |    ATN       |
        B_PBShi: byte;                    //       PS|

        G_VMS_KMOD: byte;                 //       0 |    |    |    | VMS|   KMOD   |
        G_NU33: byte;
        G_PerfName01: byte;               //       ASCII
        G_PerfName02: byte;               //       ASCII
        G_PerfName03: byte;               //       ASCII
        G_PerfName04: byte;               //       ASCII
        G_PerfName05: byte;               //       ASCII
        G_PerfName06: byte;               //       ASCII
        G_PerfName07: byte;               //       ASCII
        G_PerfName08: byte;               //       ASCII
        G_PerfName09: byte;               //       ASCII
        G_PerfName10: byte;               //       ASCII
        G_PerfName11: byte;               //       ASCII
        G_PerfName12: byte;               //       ASCII
        G_PerfName13: byte;               //       ASCII
        G_PerfName14: byte;               //       ASCII
        G_PerfName15: byte;               //       ASCII
        G_PerfName16: byte;               //       ASCII
        G_PerfName17: byte;               //       ASCII
        G_PerfName18: byte;               //       ASCII
        G_PerfName19: byte;               //       ASCII
        G_PerfName20: byte;               //       ASCII
        G_PerfName21: byte;               //       ASCII
        G_PerfName22: byte;               //       ASCII
        G_PerfName23: byte;               //       ASCII
        G_PerfName24: byte;               //       ASCII
        G_PerfName25: byte;               //       ASCII
        G_PerfName26: byte;               //       ASCII
        G_PerfName27: byte;               //       ASCII
        G_PerfName28: byte;               //       ASCII
        G_PerfName29: byte;               //       ASCII
        G_PerfName30: byte;               //       ASCII
      )
  end;
`

**Comment by BobanSpasic** [comment]
> Wow. So what do you think of the idea to be able to open performance.syx and save them as Performance.ini? By the way, we should link to MiniDexedLibrarian more prominently. Will do so in the README.

I will write a separate command line convertor. 

btw. I've also prepared own sysex format for minidexed performances if we ever implement MIDI dumps for something like that.

**Reply by probonopd** [reply]
Looking forward to the command line convertor. :+1: 
Maybe we could even build an on-the-fly translator into the MiniDexed firmware? Let's see. Rome was not built in one day.

**Reply by probonopd** [reply]
> own sysex format for minidexed performances

Let's understand the TX/DX performance sysex first; maybe we could even use that (and maybe utilize some reserved/unused bytes for our extensions)? On the other hand, I can also see the advantage of defining our own. I guess it depends on whether one wants to use existing PC software that was written for the DX/TX or whether one assumes MiniDexed specific software (imagine a Dexed-like vst that would be compatible to MiniDexed performances...)

**Reply by probonopd** [reply]
So, in order to convert DX1/DX5/TX7/... performances to MiniDexed performances, we would need to interpret both PMEM and PCED, because MiniDexed performances contain both types of information:

* **PMEM** (Performance Memory) Format: Contain performance data, which includes settings for voices, effects, and other parameters that define how the synthesizer will sound.
* **PCED** (Preset Combination Edit Data) Format: Store preset combination edit data, enabling the customization of preset voices and other settings. By modifying PCED files, you can create new voice combinations or modify existing presets.

Correct?

**Reply by BobanSpasic** [reply]
A bit of theory:
All the DX synths have an edit buffer. At the moment you change one parameter for a stored voice, the voice will be copied to the edit buffer and edited there. All the xCED formats are related to the edit buffer (VCED, PCED, ACED).
At sending a single voice from PC to the DX in VCED format - it goes to the edit buffer. You need (on the synth) to store/save it somewhere for later use.

All the xMEM formats (VMEM, AMEM, PMEM) are for sending the whole banks over the MIDI. The voices are automatically stored at the positions predefined in the sysex file.
The xMEM formats are more compact because of the "compression". It means, if some parameter is e.g. 3 bits long, and some other is 4 bits long - they are combined in one single byte for the transport.
`A_BCA_BCS: byte;                  //           BCA      |        BCS        |`
Here you see BCA taking 3 bits and BCS taking 4 bits, combined in one byte.
Don't forget that SysEx is 7bit-based. The most significant bit is never used and must be set to 0. If some parameter extends over two bytes, the result is 128 x MSB + LSB, and not 256 x MSB + LSB like we know from 8bit systems.

**Reply by probonopd** [reply]
I tried to start putting together performance `.ini` files for the DX5 performances:
https://github.com/probonopd/MiniDexed/tree/dx5_performances/performance (from `[000131_DOUBLE_HOR.ini](https://github.com/probonopd/MiniDexed/blob/dx5_performances/performance/000131_DOUBLE_HOR.ini) on) but I cannot get them to work. What am I doing wrong?

The script I used is at https://github.com/probonopd/MiniDexed/blob/dx5_performances/performance_dx5.py

**Reply by BobanSpasic** [reply]
Just a bling guess - are you taking voice data in VMEM format from the syx files?
And you try to put it in performance.ini where VCED is expected?

**Reply by probonopd** [reply]
This did the trick!

**Reply by BobanSpasic** [reply]
Use "MDX_Tool -split" to split the VMEM into separate VCEDs, crop the headers and tails, and try again

or

wait for me to write the converter

**Reply by probonopd** [reply]
It's working now. I was able to create MiniDexed performance files for the voice combinations used in the DX5. The only thing we are currently missing to reproduce the DX5 performances properly is the factory DX5 performance data (the values of the parameters).

**Comment by BobanSpasic** [comment]
DX1 Performance/Function parameters:  
[dx1smp10.pdf](https://github.com/probonopd/MiniDexed/files/13231631/dx1smp10.pdf)

source: http://www.kratzer.at/dxtech.htm

For some of the parameters, I have no clue what they should do. I also do not see the references to the voices used in the performance.
Anyone with some experience with DX1/DX5?

**Reply by BobanSpasic** [reply]
So, I have the complete PCED format for DX1/DX5/TX7, but that won't help much without PMEM format.

``` 
 TTX7_PCED_Params = packed record
    case boolean of
      True: (params: TTX7_PCED_Dump);
      False: (
        A_NU0: byte;                      //       Not used
        A_Source_Select: byte;            //DX5    0-15 (MIDI channel)
        A_PolyMono: byte;                 //       0-1
        A_PitchBendRange: byte;           //       0-12
        A_PitchBendStep: byte;            //       0-12
        A_PortamentoTime: byte;           //       0-99
        A_PortaGlissando: byte;           //       0-1
        A_PortamentoMode: byte;           //       0-1
        A_PortamentoPedal: byte;          //DX5    0-1
        A_ModWheelSens: byte;             //       0-15  translates to 0-99 for DX7
        A_ModWheelAssign: byte;           //       0-7
        A_FootCtrlSens: byte;             //       0-15  translates to 0-99 for DX7
        A_FootCtrlAssign: byte;           //       0-7
        A_AfterTouchSens: byte;           //       0-15  translates to 0-99 for DX7
        A_AfterTouchAssign: byte;         //       0-7
        A_BrthCtrlSens: byte;             //       0-15  translates to 0-99 for DX7
        A_BrthCtrlAssign: byte;           //       0-7
        A_KeyIndividualAftTchSens: byte;  //DX5    0-15
        A_KIAT_OP1_Sens: byte;            //DX5    0-15
        A_KIAT_OP2_Sens: byte;            //DX5    0-15
        A_KIAT_OP3_Sens: byte;            //DX5    0-15
        A_KIAT_OP4_Sens: byte;            //DX5    0-15
        A_KIAT_OP5_Sens: byte;            //DX5    0-15
        A_KIAT_OP6_Sens: byte;            //DX5    0-15
        A_KIAT_DecayRate: byte;           //DX5    0-99
        A_KIAT_ReleaseRate: byte;         //DX5    0-99
        A_VoiceAttn: byte;                //       0-7  translates to 0-99 for DX7
        A_ProgramOutput: byte;            //DX5    0-1
        A_SustainPedal: byte;             //DX5    0-1
        A_PerfKeyShift: byte;             //DX5    0-48  center=24

        B_NU0: byte;                      //       Not used
        B_Source_Select: byte;            //DX5    0-15 (MIDI channel)
        B_PolyMono: byte;                 //       0-1
        B_PitchBendRange: byte;           //       0-12
        B_PitchBendStep: byte;            //       0-12
        B_PortamentoTime: byte;           //       0-99
        B_PortaGlissando: byte;           //       0-1
        B_PortamentoMode: byte;           //       0-1
        B_PortamentoPedal: byte;          //DX5    0-1
        B_ModWheelSens: byte;             //       0-15  translates to 0-99 for DX7
        B_ModWheelAssign: byte;           //       0-7
        B_FootCtrlSens: byte;             //       0-15  translates to 0-99 for DX7
        B_FootCtrlAssign: byte;           //       0-7
        B_AfterTouchSens: byte;           //       0-15  translates to 0-99 for DX7
        B_AfterTouchAssign: byte;         //       0-7
        B_BrthCtrlSens: byte;             //       0-15  translates to 0-99 for DX7
        B_BrthCtrlAssign: byte;           //       0-7
        B_KeyIndividualAftTchSens: byte;  //DX5    0-15
        B_KIAT_OP1_Sens: byte;            //DX5    0-15
        B_KIAT_OP2_Sens: byte;            //DX5    0-15
        B_KIAT_OP3_Sens: byte;            //DX5    0-15
        B_KIAT_OP4_Sens: byte;            //DX5    0-15
        B_KIAT_OP5_Sens: byte;            //DX5    0-15
        B_KIAT_OP6_Sens: byte;            //DX5    0-15
        B_KIAT_DecayRate: byte;           //DX5    0-99
        B_KIAT_ReleaseRate: byte;         //DX5    0-99
        B_VoiceAttn: byte;                //       0-7  translates to 0-99 for DX7
        B_ProgramOutput: byte;            //DX5    0-1
        B_SustainPedal: byte;             //DX5    0-1
        B_PerfKeyShift: byte;             //DX5    0-48  center=24

        G_KeyAssignMode: byte;            //DX5    0-2   0,1,2 = single, dual, split
        G_VoiceMemSelFlag: byte;          //       0-1   if KMOD=0: VMS=0 > Voice A plays; VMS=1 Voice B plays
        G_DualModeDetune: byte;           //DX5    0-15
        G_SplitPoint: byte;               //DX5    0-99
        G_PerfName01: byte;               //       ASCII
        G_PerfName02: byte;               //       ASCII
        G_PerfName03: byte;               //       ASCII
        G_PerfName04: byte;               //       ASCII
        G_PerfName05: byte;               //       ASCII
        G_PerfName06: byte;               //       ASCII
        G_PerfName07: byte;               //       ASCII
        G_PerfName08: byte;               //       ASCII
        G_PerfName09: byte;               //       ASCII
        G_PerfName10: byte;               //       ASCII
        G_PerfName11: byte;               //       ASCII
        G_PerfName12: byte;               //       ASCII
        G_PerfName13: byte;               //       ASCII
        G_PerfName14: byte;               //       ASCII
        G_PerfName15: byte;               //       ASCII
        G_PerfName16: byte;               //       ASCII
        G_PerfName17: byte;               //       ASCII
        G_PerfName18: byte;               //       ASCII
        G_PerfName19: byte;               //       ASCII
        G_PerfName20: byte;               //       ASCII
        G_PerfName21: byte;               //       ASCII
        G_PerfName22: byte;               //       ASCII
        G_PerfName23: byte;               //       ASCII
        G_PerfName24: byte;               //       ASCII
        G_PerfName25: byte;               //       ASCII
        G_PerfName26: byte;               //       ASCII
        G_PerfName27: byte;               //       ASCII
        G_PerfName28: byte;               //       ASCII
        G_PerfName29: byte;               //       ASCII
        G_PerfName30: byte;               //       ASCII
      )
  end;
```

**Comment by pik33** [comment]
When I experimented with the PC Dexed, there was a big library of .syx files including folder named "original yamaha". I don't know how "original" it is but here it is
[Original Yamaha.zip](https://github.com/probonopd/MiniDexed/files/13259944/Original.Yamaha.zip)

**Reply by probonopd** [reply]
Thanks, but these are voice, not performance files.

**Comment by BobanSpasic** [comment]
@probonopd 
I am almost finished with the converter, but... the DX5 files that you've provided in https://github.com/probonopd/MiniDexed/discussions/554#discussioncomment-7448900 are corrupted. 
The Performance file is corrupted somewhere in the middle, a couple of performances are missing. The Voice files have a wrong header (this can be easily repaired).
Any other copies available?

**Reply by probonopd** [reply]
Ah, I was also stumbling on this in my Python script, thought I had made some mistake. 
Unfortunately so far I know only the DX5 performance file linked at https://yamahamusicians.com/forum/viewtopic.php?p=114756#p114756, which seems to be not only corrupted but also contain other performances than the factory ones. Maybe we can track down the DX5 factory performances sysex somewhere, so far I have not found it.

**Comment by BobanSpasic** [comment]
[DX7II.ZIP](https://github.com/probonopd/MiniDexed/files/13301521/DX7II.ZIP)  (edit: don't download, full of errors)
[MDX_PerfConv.zip](https://github.com/probonopd/MiniDexed/files/13301522/MDX_PerfConv.zip)  (edit: don't download, full of errors)

I am having trouble with setting up a Github repository for the converter. Here is the upload of the Windows 64bit version of MDX_PerfConv, and some files converted from DX7II files from bladeSk: https://github.com/bladeSk/DX7II-Librarian/issues/3
I can't test the converted files at the moment. Please give it a try and report issues.

**Reply by probonopd** [reply]
Hi @BobanSpasic, thanks for MDX_PerfConv. Unfortunately the converted performances don't sound quite right, although I wasn't able to analyze yet why. Maybe you have a chance to test with a RPi?

**Reply by BobanSpasic** [reply]
Yeah, I forgot to turn off the unused TGs, so these were playing init sound because MIDI Ch was set to omni.
btw. what are the rules for file names? From 32 performances, my MiniDexed sees just 8 performances. Is there a problem with file name length, or file size, or chars that are not allowed?

Next try:
[DX7II.ZIP](https://github.com/probonopd/MiniDexed/files/13321620/DX7II.ZIP)   (edit: don't download, file names too long)

**Reply by BobanSpasic** [reply]
Answering to my self - max length of the performance name is 14 (+6 for performance number + 1 for underscore = 21)

**Reply by probonopd** [reply]
Let's document performance.ini more at https://github.com/probonopd/MiniDexed/wiki/Files#performanceini :-)

**Reply by BobanSpasic** [reply]
Tonight on the program for the converter:
- tune the Detune parameter. DX7II's Dual detune does a +-25 cent for the 0-15 parameter values. DX1 does +-25 cent for 32 parameter values (double the precision of the DX7II). MiniDexed have -99 to 99 as the parameters, but how this translates to cents?
- check how the transpose is working on MiniDexed. DX7II transposes two times - sum the transpose in voice parameters and the transpose from performance parameters

**Reply by probonopd** [reply]
And maybe use the DXII factory performances for testing (rather than the DX5 ones someone had converted to the DXII format). Probably there are some DXII features which MiniDexed (and Dexed?) don't support yet, such as Microtuning?

**Reply by BobanSpasic** [reply]
I've found some DX7II banks with performances and I got into another question.
DX7II has a couple of controllers (wheels, pedals, sliders etc.) and these performances are using them. How to handle the conversion?
Should I generate a separate file with these parameters as comments, or should I extend the performance.ini format (even if MiniDexed does not make use of the extensions at the moment)?

**Reply by probonopd** [reply]
Great point. If MiniDexed performance files support comments, I'd put in something into the performance files like

```
 # TODO: SomeKeyThatDoesNotExistYet=Value
```

This way, we know what we need to work on :-)

**Comment by BobanSpasic** [comment]
[DX7II.ZIP](https://github.com/probonopd/MiniDexed/files/13312808/DX7II.ZIP)   (edit: don't download, full of errors)
Both performance banks converted.

MDX_PerfConv is now located here: https://github.com/BobanSpasic/MDX_PerfConv
